contracts/zero-ex/contracts/src/features/UniswapFeature.sol

```diff
@@ -0,0 +1,320 @@
+                mstore(0xB44, receiver)
+                mstore(0xB64, 0x80)
+                mstore(0xB84, 0)
+                if iszero(call(gas(), pair, 0, 0xB00, 0xA4, 0, 0)) {
+                    bubbleRevert()
+                }
+            } // End for-loop.
+
+            // If buying ETH, unwrap the WETH first
+            if eq(buyToken, ETH_TOKEN_ADDRESS_32) {
+                // Call `WETH.withdraw(buyAmount)`
+                mstore(0xB00, WETH_WITHDRAW_CALL_SELECTOR_32)
+                mstore(0xB04, buyAmount)
+                if iszero(call(gas(), mload(0xA40), 0, 0xB00, 0x24, 0x00, 0x0)) {
+                    bubbleRevert()
+                }
+                // Transfer ETH to the caller.
+                if iszero(call(2300, caller(), buyAmount, 0xB00, 0x0, 0x00, 0x0)) {
+                    bubbleRevert()
+                }
+            }
+
+            // Functions ///////////////////////////////////////////////////////
+
+            function loadTokenAddress(idx) -> addr {
+                addr := and(ADDRESS_MASK, calldataload(add(mload(0xA00), mul(idx, 0x20))))
+            }
+
+            function normalizeToken(token) -> normalized {
+                normalized := token
+                // Translate ETH pseudo-tokens to WETH.
+                if eq(token, ETH_TOKEN_ADDRESS_32) {
+                    normalized := mload(0xA40)
+                }
+            }
+
+            function computePairAddress(tokenA, tokenB) -> pair {
+                // Compute the UniswapV2Pair address
+                tokenA := normalizeToken(tokenA)
+                tokenB := normalizeToken(tokenB)
+                // Correct order of tokens.
+                switch lt(tokenA, tokenB)
+                    case 0 {
+                        mstore(0xB14, tokenA)
+                        mstore(0xB00, tokenB)
+                    }
+                    default {
+                        mstore(0xB14, tokenB)
+                        mstore(0xB00, tokenA)
+                    }
+                let salt := keccak256(0xB0C, 0x28)
+                switch mload(0xA20) // isSushi
+                    case 0 {
+                        mstore(0xB00, FF_UNISWAP_FACTORY)
+                        mstore(0xB15, salt)
+                        mstore(0xB35, UNISWAP_PAIR_INIT_CODE_HASH)
+                    }
+                    default {
+                        mstore(0xB00, FF_SUSHISWAP_FACTORY)
+                        mstore(0xB15, salt)
+                        mstore(0xB35, SUSHISWAP_PAIR_INIT_CODE_HASH)
+                    }
+                pair := and(ADDRESS_MASK, keccak256(0xB00, 0x55))
+            }
+
+            function bubbleRevert() {
+                returndatacopy(0, 0, returndatasize())
+                revert(0, returndatasize())
+            }
+        }
+
+        // Revert if we bought too little.
+        // TODO: replace with rich revert?
+        require(buyAmount >= minBuyAmount, "UniswapFeature/UnderBought");
```

Comment:
If there are bugs, the check here against `minBuyAmount` can give people confidence that the outcome will be correct. That said, there's no explicit guarantee in this code that the requested amount was actually delivered to the sender. We're relying on Uniswap's implementation (which _does_ have a fairly explicit guarantee). I think this is okay, but it feels like the biggest risk. I haven't been able to think of a better check.